### PR TITLE
Refactor `CannotMakeMoreVisible`

### PR DESCRIPTION
**Pull request by dougbu**
- #8410
- handle the new combinations now possible w/ `--respect-internals`

### CODE REVIEW

**Comment by dougbu** [comment]
I may be missing corner cases and would appreciate eyes on this before we think about adding tests for the #8410 scenario.

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 1]
```diff
@@ -16,19 +16,11 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinitionMem
...
 
-            // If implementation is public then contract can be any visibility
-            if (impl.Visibility == TypeMemberVisibility.Public)
-                return DifferenceType.Unknown;
-
-            // If implementation is protected or protected internal then contract must be protected or protected internal as well.
-            if (impl.Visibility == TypeMemberVisibility.Family || impl.Visibility == TypeMemberVisibility.FamilyOrAssembly)
+            if (HasReducedVisibility(contract.Visibility, impl.Visibility))
             {
-                if (contract.Visibility != TypeMemberVisibility.Family && contract.Visibility != TypeMemberVisibility.FamilyOrAssembly)
```
Comment (dougbu) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:26]: Logic like this was incorrect for the `--respect-internals` case because `protected` is a downgrade from `protected internal` as far as friend assemblies are concerned.

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 2]
```diff
@@ -39,22 +31,44 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                case TypeMemberVisibility.Public:
+                    // If implementation is public then contract can be any visibility.
+                    return false;
+                case TypeMemberVisibility.FamilyOrAssembly:
+                    // protected internal is an upgrade from everything but public.
+                    return contract == TypeMemberVisibility.Public;
+                case TypeMemberVisibility.Assembly:
+                    // internal is an upgrade only from private or private protected.
+                    return contract != TypeMemberVisibility.Private && contract != TypeMemberVisibility.FamilyAndAssembly;
+                case TypeMemberVisibility.Family: // protected
```
Comment (dougbu) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:62]: I'm least sure about this point. Wondering if `protected` is also a compatible upgrade from `private protected` (just in a different direction than `internal`)&hellip;

**Comment by dougbu** [comment]
@ericstj @safern I rebased on 'main' and addressed my own concern about `protected` accessibility. Please let me know what you think before I add a few tests.

**Comment by dougbu** [comment]
@ViktorHofer and @wtgodbe welcome to the party 😃 I invited you because you have some history in this codebase.

@jaredpar I'd appreciate your thoughts too (of course). If there's someone on your team you particularly likes accessibility modifiers, the more the merrier.

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 3]
```diff
@@ -39,22 +30,44 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                return false;
+            }
+
+            switch (implementation)
+            {
+                case TypeMemberVisibility.Public:
+                    // If implementation is public then contract can be any visibility.
+                    return false;
+                case TypeMemberVisibility.FamilyOrAssembly:
+                    // protected internal is an upgrade from everything but public.
```
Comment (ericstj) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:56]: nit: `upgrade` here had me a bit confused.  Is it possible to word these in terms of `reduced` since that's the method name?

**Comment by ericstj** [comment]
The change looks reasonable to me modulo a small nit on the comments.  Tests would be good, also a trial build of dotnet/runtime and/or aspnetcore to make sure nothing regressed.

**Comment by dougbu** [comment]
> modulo a small nit on the comments

I'll address that suggestion as I add tests.

> a trial build of dotnet/runtime and/or aspnetcore to make sure nothing regressed

This is possible for dotnet/runtime but we never switched dotnet/aspnetcore to use ApiCompat. I'll do it w/ dotnet/runtime offline.

**Comment by dougbu** [comment]
Separately, thoughts on `CannotMakeMoreVisible` to `CannotMakeLessVisible`❔ The new name would be more correct but I'm not sure about the specific compatibility requirements here.

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs (hunk 1 above)
Comment (dougbu) [reply src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:26]: This class doesn't know whether `--respect-internals` was specified. The new implementation is a slight breaking change 
 when `protected` and `protected internal` are used for the same member or type in the contract and implementation or visa versa. Is that a concern❔

**Comment by ericstj** [comment]
> The new name would be more correct but I'm not sure about the specific compatibility requirements here.

Feel free to rename to whatever is clearest.  These rules are only used internally.

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs (hunk 3 above)
Comment (bartonjs) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:56]:
```suggestion
                    // `protected internal` more visible than everything but public.
```

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 4]
```diff
@@ -39,22 +30,44 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                    return contract == TypeMemberVisibility.Public;
+                case TypeMemberVisibility.Assembly: // internal
+                case TypeMemberVisibility.Family: // protected
+                    // internal and protected are upgrades only from private or private protected.
```
Comment (bartonjs) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:60]:
```suggestion
                    // `private` or `private protected` are the only things less visible than either `internal` or `protected`
```

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 5]
```diff
@@ -39,22 +30,44 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                    return contract != TypeMemberVisibility.Private && contract != TypeMemberVisibility.FamilyAndAssembly;
+                case TypeMemberVisibility.FamilyAndAssembly:
+                    // private protected is very restrictive; only an upgrade from private.
```
Comment (bartonjs) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:63]:
```suggestion
                    // `private` is the only scope less visible than `private protected`
```

File: src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs [hunk 6]
```diff
@@ -39,22 +30,44 @@ public override DifferenceType Diff(IDifferences differences, ITypeDefinition im
...
+                    return contract != TypeMemberVisibility.Private;
+                case TypeMemberVisibility.Private:
+                    // private in the implementation is always a downgrade.
```
Comment (bartonjs) [inline src/Microsoft.DotNet.ApiCompat/src/Rules/Compat/CannotMakeMoreVisible.cs:66]:
```suggestion
                    // `private` is the least visible scope, so it's automatically less visible than the contract indicated.
```

**Comment by bartonjs** [comment]
I gave some suggested rewordings to the comments to avoid "upgrade" and "downgrade", but instead to always define the state as "more visible" or "less visible".

**Comment by dougbu** [comment]
> Feel free to rename to whatever is clearest.

👍 @ericstj 

> I gave some suggested rewordings to the comments to avoid "upgrade" and "downgrade", but instead to always define the state as "more visible" or "less visible".

Much appreciated @bartonjs❕ I made some changes to the comments locally but keep getting pulled away from finishing work on testing these changes. I'll incorporate your suggestions (which I like better than my updates) in my next iteration.

**Comment by dougbu** [comment]
no longer required
